File: backend/src/main/java/com/example/backend/model/Hotel.java
package com.example.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "hotels")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Hotel {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "hotel_id")
    private UUID id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "city", nullable = false, length = 100)
    private String city;

    @Column(name = "address")
    private String address;

    @Column(name = "rating")
    private Double rating;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "thumbnail_url")
    private String thumbnailUrl;

    @Column(name = "images")
    private List<String> images;

    @Column(name = "latitude")
    private Double latitude;

    @Column(name = "longitude")
    private Double longitude;

    @Column(name = "contact_phone", length = 20)
    private String contactPhone;

    @Column(name = "contact_email", length = 100)
    private String contactEmail;

    @Column(name = "check_in_time")
    private String checkInTime;

    @Column(name = "check_out_time")
    private String checkOutTime;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // N: 1 relationship with User
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "manager_id", nullable = false)
    private User manager;

    // 1: N relationship with RoomType
    @OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RoomType> roomTypes = new ArrayList<>();

    // 1: N relationship with Review
    @OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Review> reviews;
}

File: frontend/src/types/hotel.ts
import type { BaseEntity, BaseFilterRequest } from "./common";
import type { ReviewResponse } from "./review";
import type { RoomTypeResponse } from "./roomType";
import type { ManagerResponse } from "./user";

export interface HotelListResponse extends BaseEntity {
  name: string;
  city: string;
  address: string;
  rating: number;
  description: string;
  thumbnailUrl: string;
  checkInTime: string;
  checkOutTime: string;
}

export interface HotelDetailResponse extends BaseEntity {
  name: string;
  city: string;
  address: string;
  rating: number;
  description: string;
  thumbnailUrl: string;
  images: string[];
  manager: ManagerResponse;
  latitude: number;
  longitude: number;
  contactPhone: string;
  contactEmail: string;
  checkInTime: string;
  checkOutTime: string;
  roomTypes: RoomTypeResponse[];
  reviews: ReviewResponse[];
}

export interface HotelRequest {
  name: string;
  city: string;
  address: string;
  rating?: number;
  description?: string;
  thumbnailUrl?: string;
  images?: string[];
  managerId: string;
  latitude?: number;
  longitude?: number;
  contactPhone: string;
  contactEmail: string;
  checkInTime?: string;
  checkOutTime?: string;
}

export interface HotelUpdateRequest {
  name?: string;
  city?: string;
  address?: string;
  rating?: number;
  description?: string;
  managerId?: string;
  latitude?: number;
  longitude?: number;
  contactPhone?: string;
  contactEmail?: string;
  checkInTime?: string;
  checkOutTime?: string;
  thumbnailUrl?: string;
  images?: string[];
}

export interface HotelFilterRequest extends BaseFilterRequest {
  city?: string;
  name?: string;
  address?: string;
  keyword?: string;
  managerId?: string;
  minRating?: number;
  maxRating?: number;
  guests?: number;
  checkIn?: string;
  checkOut?: string;
}
